src/clock.cr

```diff
@@ -0,0 +1,103 @@
+# Monotonic Clock.
+#
+# A monotonic clock is tied to the system and isn't affected by time
+# fluctuations, for example leap seconds or manually changing the computer's
+# time.
+#
+# Compared to `Time`, the `Clock` value isn't interesting and most likely means
+# nothing, but its value is guaranteed to be linearily increasing from a fixed
+# point in time (whose origin is unspecified), thus useful to calculate the
+# elapsed time between two clocks.
+#
+# Example:
+#
+# ```
+# timeout = 5.seconds
+# clock = Clock.monotonic
+#
+# until clock.elapsed?(timeout)
+#   do_domething
+# end
+# ```
+struct Clock
+  include Comparable(self)
+  include Comparable(Time::Span)
+  include Comparable(Float64 | Float32)
+
+  getter seconds : Float64
+
+  {% if flag?(:darwin) %}
+    @@mach_timebase_info : LibC::MachTimebaseInfo?
+
+    private def self.mach_timebase_info
+      @@mach_timebase_info ||= begin
+        LibC.mach_timebase_info(pointerof(out info))
+        info
+      end
+    end
+  {% end %}
+
+  def self.monotonic : self
+    {% if flag?(:darwin) %}
```

Comment:
Note that Sierra now supports `clock_gettime`, but since the precision isn't as good, and more importantly it's not portable (a binary won't run on older versions) I chose to stick to `mach_absolute_time`.